### PR TITLE
Introduce http-client + http-data

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview

This update continues development of the error-reporting feature and adds a middle-ware piece that provides an API the game-engine client can use to communicate to the http-server. The http client-server interchange format for communication is JSON formatted String which can then be automatically converted to Java objects pretty easily.

The API is developed to a nearly production ready detail, the feature for error-reporting is expanded to support user input to describe the problem to us, which opens up the feature for ad-hoc error-reporting even without a stack trace (so more than just an automated stack trace upload).

Http-client would ideally provide as much validation and safety checks as we can, the entry points are minimal and that helps then guarantee the integrity of the rest of the system. To this extent, the client has a built in throttle to drop messages that wind up being too large (a large arbitrary number was chosen to start to give us an upper bound), and a simplistic rate throttle so only one error report can be sent every few seconds.

The http-client is tested with '[WireMock](http://wiremock.org/)' which is a stand-alone http server simulator (so this update is independent of http-server). Using that we can test failure scenarios very closely.

To avoid re-inventing the wheel http client, [Feign](https://www.baeldung.com/intro-to-feign) is picked up to create an http client implementation that is bound to an API that we define as an interface with annotations. This reduces the need for unit testing to a pretty reasonable minimum as largely we just define an interface and then use a client builder to create the concrete implementation. The burden is then largely for system/integration testing to verify that things would work provided we use the API presented by http-client sub-project.

Next steps:
 - game engine client code update to use the new http-client and have the error reporting feature behind a feature flag.
 - server side updates to comply with the API now defined by the client.
 - infrastructure updates to stand-up the http server

## Functional Changes
- none, two new libraries added: http-client + http-data, but neither are used by prod code yet.
- when triplea game-core does include http-client as a dependnecy, the new Feign libraries for http client will add 100KB to the DL size
- there is a refactoring change in the main game code to receive a LogRecord instead of 'String'. This is done so that when we tie into the http-client we will have access to the exception in the LogRecord and could then send it.

## Manual Testing Performed
- none

### After
screenshot from 2018-09-03 15-37-57

### CODE REVIEW

File: http-client/src/main/java/org/triplea/http/client/ErrorReportingHttpClient.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package org.triplea.http.client;
```
Comment (ssoloff) [inline http-client/src/main/java/org/triplea/http/client/ErrorReportingHttpClient.java]: :+1: Seems like you've standardized the package names in these two new projects.  Will there be a follow-up PR to update the `http-server` project to use `org.triplea.http.server` as the root package?

File: game-core/src/main/java/games/strategy/debug/ErrorMessage.java [hunk 2]
```diff
@@ -92,10 +93,10 @@ public static void enable() {
     INSTANCE.enableErrorPopup = true;
   }
 
-  public static void show(final String msg) {
+  public static void show(final LogRecord record) {
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/debug/ErrorMessage.java]: Is this where you plan on initiating the report to the HTTP server?  If so, we may want to reconsider that.  This class is a UI component, and it probably shouldn't have responsibility for sending stuff over the network.

We now have three sinks for logs:

1. The console
1. The `ErrorMessage` UI
1. The HTTP error reporting service

Maybe we should take a step back and install multiple handlers into the `Logger` for each of these sinks instead of having everything piggyback off of `ConsoleHandler`?

File: http-client/src/main/java/org/triplea/http/client/ErrorReportingHttpClient.java (hunk 1 above)
Comment (DanVanAtta) [reply http-client/src/main/java/org/triplea/http/client/ErrorReportingHttpClient.java]: Good question, I'm not sure. I'm open to suggestions on this one. Here is how it looks for me:

screenshot from 2018-09-03 19-56-33

In the above, I like how `http` splits the http server pieces away from the business/domain processing. Anything not in the `http` layer will have no dependency on that, http talks to `reporting.error` by passing it objects from `http-data` (but in this case the API is still a `String`, it's still yet to be updated and is the simplistic initial server API)

Another way to look at it, I don't want `org.triplea.server.reporting.error.upload` to have `http` in it anywhere:
- would couple to one server implementation or another, so replacing spark with dropwizard might not be simple if we did that
- help keep some business layer processing purity; keep the processing domain split from http handling.


Should:  `org.triplea.server.reporting.error.upload`  be updated in terms of package name?
Second question, where would the `http.spark` and `http.dropwizard` branches start?

File: game-core/src/main/java/games/strategy/debug/ErrorMessage.java (hunk 2 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/ErrorMessage.java]: In one manner, but not quite as a stream. For a first and second rev having something that can be used as a generic error reporting tool independent of an error seems useful. We can attach some nice functionality to that like upload save game and screenshots, but fundamentally the upload is user driven and they can attach repro/description text to the error message.

In this location the intent was to add a button that would pass the `LogRecord` data to it's action handler if user chooses a 3rd option instead of closing an error or viewing details, they can upload.

File: http-client/src/main/java/org/triplea/http/client/throttle/rate/RateLimitingThrottle.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    this.minMillisBetweenRequests = minMillisBetweenRequests;
+    this.instantSupplier = instantSupplier;
+  }
+
+  @Override
+  public void accept(final ErrorReport errorReport) {
+    final Instant nextInstant = instantSupplier.get();
+
+    final long millisSinceLast;
+    millisSinceLast = lastInstant.until(nextInstant, ChronoUnit.MILLIS);
```
Comment (ssoloff) [inline http-client/src/main/java/org/triplea/http/client/throttle/rate/RateLimitingThrottle.java]: Can these two lines be merged?

File: http-client/src/test/java/org/triplea/http/client/throttle/size/MessageSizeThrottleTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @SuppressWarnings("JUnitTestMethodWithNoAssertions")
+  @Test
+  void messageSizeUnderLimitIsAllowed() {
+    final int characterLimit = ERROR_REPORT.toString().length() + 1;
+
+    new MessageSizeThrottle(characterLimit)
+        .accept(ERROR_REPORT);
+    // no exceptions expected
```
Comment (ssoloff) [inline http-client/src/test/java/org/triplea/http/client/throttle/size/MessageSizeThrottleTest.java]: You can use `org.triplea.test.common.Assertions.assertNotThrows()` to document this instead of the comment (and remove the `@SuppressWarnings` at the same time).

File: http-data/build.gradle [hunk 5]
```diff
@@ -0,0 +1,5 @@
+description = 'Contains common data models exchanged between client/server'
+
+dependencies {
+
+}
```
Comment (ssoloff) [inline http-data/build.gradle]: Does Gradle require this?  Can it just be removed?

Comment (DanVanAtta) [reply http-data/build.gradle]: Seems not required, removed the unnecessary boiler plate in: 03cdaaa

File: game-core/src/main/java/games/strategy/debug/ErrorMessage.java (hunk 2 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/ErrorMessage.java]: No worries, I think the original thinking was a behind the scenes error reporter. A stream of errors may not be that useful, attaching user-messages will help the quality of the error report and also allows us to have a generic in-game reporting feature. That was a pretty late-breaking shift in the intended feature.

File: http-client/src/test/java/org/triplea/http/client/throttle/size/MessageSizeThrottleTest.java (hunk 4 above)
Comment (ssoloff) [reply http-client/src/test/java/org/triplea/http/client/throttle/size/MessageSizeThrottleTest.java]: Is the `@SuppressWarnings("JUnitTestMethodWithNoAssertions")` still required on the enclosing method after the latest change to use `assertNotThrows()`?
